app/src/main/java/com/youssef/fixit/Auth/LoginFragment/LoginPresenter.java

```diff
@@ -0,0 +1,108 @@
+package com.youssef.fixit.Auth.LoginFragment;
+
+import android.content.Intent;
+import android.util.Log;
+import android.widget.Toast;
+
+import com.youssef.fixit.MainActivity.MainActivity;
+import com.youssef.fixit.Models.Bids.CreateBid;
+import com.youssef.fixit.Models.Data.RetrofitClient;
+import com.youssef.fixit.Models.Register.Register;
+
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+class LoginPresenter {
+    LoginView view;
+
+    public LoginPresenter(LoginView view) {
+        this.view = view;
+    }
+
+    void OnLogin(String Mail, String Password) {
+        Validation(Mail, Password);
+    }
+
+    void Validation(String Mail, String Password) {
+        if (Mail.isEmpty()) {
+            view.OnMailIsError("Please enter your Mail");
+            return;
+        }
+        if (Password.isEmpty()) {
+            view.OnPasswordIsError("Please enter your Password");
+            return;
+        }
+
+        LoginRequest(Mail, Password);
+    }
+
+    void LoginRequest(String Mail, String Password) {
+        RetrofitClient.getInstance().Login(Mail, Password).enqueue(new Callback<Register>() {
+            @Override
+            public void onResponse(Call<Register> call, Response<Register> response) {
```

Comment:
A side note to be discussed later, network layer here can't be mocked, so we will need to move it to another manager class.

Ignore it for now we will discuss it when we discuss the dependency injection and dependency inversion

Reply:
What do you think if move it to model class?
i think it will achieve  single responsibility principle

Reply:
Would be better yes. you can move it to another manager then inject it in loginPresenter constructor 
So it will be something like that

```
class LoginManager() { 
public void login(LoginCallback loginCallback) {
// do the network request here and the result will be passed through the callback.
}
}
```
Then in the `loginPresenter` you will pass `LoginManager` to it via constructor or setter method.

This concept called  [Dependency Injection](https://medium.com/groupon-eng/dependency-injection-in-java-9e9438aa55ae) 

Please, don't address this comment here. We will work on the dependency injection in another pr.
For now just address the other comments I left and then after merging this branch will work on dependency inject and dependency inversion.